--- a/Src/init.c
+++ b/Src/init.c
@@ -1538,7 +1538,7 @@
     char *h;
 
     queue_signals();
-    if (EMULATION(EMULATE_SH|EMULATE_KSH) || !(h = getsparam_u("ZDOTDIR"))) {
+    if (EMULATION(EMULATE_SH|EMULATE_KSH) || !(h = getsparam("ZDOTDIR"))) {
 	h = home;
 	if (!h) {
 	    unqueue_signals();
--- a/Src/subst.c
+++ b/Src/subst.c
@@ -256,6 +256,8 @@
 		return NULL;
 	    if (!subst)
 		rest = subst = "";
+	    else
+		subst = metafy(subst, -1, META_HEAPDUP);
 
 	    sublen = strlen(subst);
 	    restlen = strlen(rest);
--- a/Src/Modules/system.c
+++ b/Src/Modules/system.c
@@ -368,9 +368,9 @@
     }
 
     if (flags & O_CREAT)
-	fd = open(*args, flags, perms);
+	fd = open(unmeta(*args), flags, perms);
     else
-	fd = open(*args, flags);
+	fd = open(unmeta(*args), flags);
 
     if (fd == -1) {
 	zwarnnam(nam, "can't open file %s: %e", *args, errno);
--- a/Src/parse.c
+++ b/Src/parse.c
@@ -3186,7 +3186,7 @@
 	    zwarnnam(nam, "too few arguments");
 	    return 1;
 	}
-	if (!(f = load_dump_header(nam, (strsfx(FD_EXT, *args) ? *args :
+	if (!(f = load_dump_header(nam, unmeta(strsfx(FD_EXT, *args) ? *args :
 					 dyncat(*args, FD_EXT)), 1)))
 		return 1;
 
@@ -3386,8 +3386,8 @@
     if (!strsfx(FD_EXT, dump))
 	dump = dyncat(dump, FD_EXT);
 
-    unlink(dump);
-    if ((dfd = open(dump, O_WRONLY|O_CREAT, 0444)) < 0) {
+    unlink(unmeta(dump));
+    if ((dfd = open(unmeta(dump), O_WRONLY|O_CREAT, 0444)) < 0) {
 	zwarnnam(nam, "can't write zwc file: %s", dump);
 	return 1;
     }
@@ -3404,7 +3404,7 @@
 	    flags = (flags & ~(FDHF_KSHLOAD | FDHF_ZSHLOAD)) | FDHF_ZSHLOAD;
 	    continue;
 	}
-	if ((fd = open(*files, O_RDONLY)) < 0 ||
+	if ((fd = open(unmeta(*files), O_RDONLY)) < 0 ||
 	    fstat(fd, &st) != 0 || !S_ISREG(st.st_mode) ||
 	    (flen = lseek(fd, 0, 2)) == -1) {
 	    if (fd >= 0)
@@ -3412,7 +3412,7 @@
 	    close(dfd);
 	    zwarnnam(nam, "can't open file: %s", *files);
 	    noaliases = ona;
-	    unlink(dump);
+	    unlink(unmeta(dump));
 	    return 1;
 	}
 	file = (char *) zalloc(flen + 1);
@@ -3424,7 +3424,7 @@
 	    zfree(file, flen);
 	    zwarnnam(nam, "can't read file: %s", *files);
 	    noaliases = ona;
-	    unlink(dump);
+	    unlink(unmeta(dump));
 	    return 1;
 	}
 	close(fd);
@@ -3436,7 +3436,7 @@
 	    zfree(file, flen);
 	    zwarnnam(nam, "can't read file: %s", *files);
 	    noaliases = ona;
-	    unlink(dump);
+	    unlink(unmeta(dump));
 	    return 1;
 	}
 	zfree(file, flen);
@@ -3522,8 +3522,8 @@
     if (!strsfx(FD_EXT, dump))
 	dump = dyncat(dump, FD_EXT);
 
-    unlink(dump);
-    if ((dfd = open(dump, O_WRONLY|O_CREAT, 0444)) < 0) {
+    unlink(unmeta(dump));
+    if ((dfd = open(unmeta(dump), O_WRONLY|O_CREAT, 0444)) < 0) {
 	zwarnnam(nam, "can't write zwc file: %s", dump);
 	return 1;
     }
@@ -3543,7 +3543,7 @@
 				 &hlen, &tlen, what)) {
 		    errflag &= ~ERRFLAG_ERROR;
 		    close(dfd);
-		    unlink(dump);
+		    unlink(unmeta(dump));
 		    return 1;
 		}
     } else if (match) {
@@ -3558,7 +3558,7 @@
 	    if (!(pprog = patcompile(pat, PAT_STATIC, NULL))) {
 		zwarnnam(nam, "bad pattern: %s", *names);
 		close(dfd);
-		unlink(dump);
+		unlink(unmeta(dump));
 		return 1;
 	    }
 	    for (i = 0; i < shfunctab->hsize; i++)
@@ -3569,7 +3569,7 @@
 				     &hlen, &tlen, what)) {
 			errflag &= ~ERRFLAG_ERROR;
 			close(dfd);
-			unlink(dump);
+			unlink(unmeta(dump));
 			return 1;
 		    }
 	}
@@ -3580,13 +3580,13 @@
 		zwarnnam(nam, "unknown function: %s", *names);
 		errflag &= ~ERRFLAG_ERROR;
 		close(dfd);
-		unlink(dump);
+		unlink(unmeta(dump));
 		return 1;
 	    }
 	    if (cur_add_func(nam, shf, lnames, progs, &hlen, &tlen, what)) {
 		errflag &= ~ERRFLAG_ERROR;
 		close(dfd);
-		unlink(dump);
+		unlink(unmeta(dump));
 		return 1;
 	    }
 	}
@@ -3595,7 +3595,7 @@
 	zwarnnam(nam, "no functions");
 	errflag &= ~ERRFLAG_ERROR;
 	close(dfd);
-	unlink(dump);
+	unlink(unmeta(dump));
 	return 1;
     }
     tlen = (tlen + hlen) * sizeof(wordcode);
@@ -3782,6 +3782,7 @@
 	tail++;
     else
 	tail = file;
+    tail = metafy(tail, -1, META_HEAPDUP);
 
     if (strsfx(FD_EXT, file)) {
 	queue_signals();
@@ -4042,7 +4043,7 @@
     if (!strsfx(FD_EXT, file))
 	file = dyncat(file, FD_EXT);
 
-    if (!(h = load_dump_header(nam, file, 1)))
+    if (!(h = load_dump_header(nam, unmeta(file), 1)))
 	return 1;
 
     for (n = firstfdhead(h), e = (FDHead) (h + fdheaderlen(h)); n < e;
